subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/FileLockCommunicator.java

```diff
@@ -68,14 +69,39 @@ public void pingOwner(int ownerPort, long lockId, String displayName) {
         } catch (IOException e) {
             throw new RuntimeException(String.format("Failed to ping owner of lock for %s (lock id: %s, port: %s)", displayName, lockId, ownerPort), e);
         }
+        return true;
     }
 
-    public long receive() throws GracefullyStoppedException {
+    public DatagramPacket receive() throws GracefullyStoppedException {
         try {
             byte[] bytes = new byte[9];
             DatagramPacket packet = new DatagramPacket(bytes, bytes.length);
             socket.receive(packet);
-            return decode(bytes);
+            return packet;
+        } catch (IOException e) {
+            if (!stopped) {
+                throw new RuntimeException(e);
+            }
+            throw new GracefullyStoppedException();
+        }
+    }
+
+    public long decodeLockId(DatagramPacket receivedPacket) {
```

Comment:
We need more information than just the decoded lock id in order to confirm ("ping back") to the requester (encoded ID, address, port).
It's then used to call `confirmUnlockRequest`: 
https://github.com/gradle/gradle/blob/5254a577d74ed5caa4e8a5d93c56d9900d5e4b2c/subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/FileLockCommunicator.java#L100-L100